Document/0x05d-Testing-Data-Storage.md

```diff
@@ -271,11 +271,104 @@ You can use stored keys in one of two modes:
+
+###### Hardware-backed Android KeyStore
+
+As mentioned before, hardware-backed Android KeyStore gives another layer to defense-in-depth security concept for Android. Keymaster Hardware Abstraction Layer (HAL) was introduced with device running Android 6 (API level 23). Application can verify if key is stored inside security hardware (by checking if `KeyInfo.isinsideSecureHardware` returns `true`). Devices running Android 9 (API level 28) and higher can have a `StrongBox Keymaster` module, an implementation of the Keymaster HAL that resides in a hardware security module which has its own CPU, Secure storage, a true random number generator and a mechanism to resist package tampering. To use this feature, `true` must be passed to the `setIsStrongBoxBacked` method in either the `KeyGenParameterSpec.Builder` class or the `KeyProtection.Builder` class when generating or importing keys using `AndroidKeystore`. To make sure that StrongBox is used during runtime check that `isInsideSecureHardware` returns `true` and that the system does not throw `StrongBoxUnavailableException` which get thrown if the StrongBox Keymaster isn't available for the given algorithm and key size associated with a key. Description of features on hardware-based keystore can be found on [AOSP pages](https://source.android.com/security/keystore "AOSP Hardware-based KeyStore").
+
+Keymaster HAL is an interface to hardware-backed component - Trusted Execution Environment (TEE) or a Secure Element (SE), which is used by Android Keystore. Example of such hardware-backed component is [Titan M](https://android-developers.googleblog.com/2018/10/building-titan-better-security-through.html "Building a Titan: Better security through a tiny chip").
+
+###### Key Attestation
+
+For the applications which heavily rely on Android Keystore for business-critical operations such as multi-factor authentication through cryptographic primitives, secure storage of sensitive data at the client-side, etc. Android provides the feature of [Key Attestation](https://developer.android.com/training/articles/security-key-attestation "Key Attestation") which helps to analyze the security of cryptographic material managed through Android Keystore. From Android 8.0 (API level 26), the key attestation was made mandatory for all new(Android 7.0 or higher) devices that need to have device certification for Google suite of apps, such devices use attestation keys signed by the [Google hardware attestation root certificate](https://developer.android.com/training/articles/security-key-attestation#root_certificate "Google Hardware Attestation Root Certificate") and the same can be verified while key attestation process.
```

Comment:
maybe good to go over the rest and see how you can use it for checkup of the security of the device> and then move that for a part into tamper detection and such...

Reply:
That leads me to consideration of some kind Android device security checklist. I think it would be good to take that  as a separate issue and define goals and objectives which we want to achieve. What do you think?

Reply:
I believe one should not be separated from another: key attestation only makes sense if you evaluate ```RootOfTrust ::= SEQUENCE {
    verifiedBootKey  OCTET_STRING,
    deviceLocked  BOOLEAN,
    verifiedBootState  VerifiedBootState,
    verifiedBootHash OCTET_STRING,
}

VerifiedBootState ::= ENUMERATED {
    Verified  (0),
    SelfSigned  (1),
    Unverified  (2),
    Failed  (3),
}```
because that will tell whether the device has been played with. This can then tell something about  what level of trust you can put into the device.

Reply:
That is clear. I added additional check which should verify device's status and simplify one of existing.